src/Plugin/resource/Resource.php

```diff
@@ -0,0 +1,303 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\restful\Plugin\resource\Resource.
+ */
+
+namespace Drupal\restful\Plugin\resource;
+
+use Drupal\Component\Plugin\PluginBase;
+use Drupal\restful\Exception\BadRequestException;
+use Drupal\restful\Exception\ForbiddenException;
+use Drupal\restful\Exception\GoneException;
+use Drupal\restful\Exception\NotImplementedException;
+use Drupal\restful\Exception\ServerConfigurationException;
+use Drupal\restful\Http\HttpHeader;
+use Drupal\restful\Http\Request;
+use Drupal\restful\Http\RequestInterface;
+use Drupal\restful\Plugin\resource\DataProvider\DataProviderInterface;
+use Drupal\restful\Plugin\resource\Field\ResourceFieldCollectionInterface;
+use Drupal\restful\Resource\ResourceManager;
+
+abstract class Resource extends PluginBase implements ResourceInterface {
+
+  /**
+   * The string that separates multiple ids.
+   */
+  const IDS_SEPARATOR = ',';
+
+  /**
+   * The requested path.
+   *
+   * @var string
+   */
+  protected $path;
+
+  /**
+   * The current request.
+   *
+   * @var RequestInterface
+   */
+  protected $request;
+
+  /**
+   * The data provider.
+   *
+   * @var DataProviderInterface
+   */
+  protected $dataProvider;
+
+  /**
+   * The field definition object.
+   *
+   * @var ResourceFieldCollectionInterface
+   */
+  protected $fieldDefinitions;
+
+  /**
+   * Data provider factory.
+   *
+   * @return DataProviderInterface
+   *   The data provider for this resource.
+   *
+   * @throws NotImplementedException
+   */
+  abstract protected function dataProviderFactory();
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRequest() {
+    if (isset($this->request)) {
+      return $this->request;
+    }
+    if (!$this->request = $this->configuration['request']) {
```

Comment:
Not sure it can happen. It's just being thorough in error handling.

Maybe for mocked classes in tests?